Decouple blockchain opcodes and state mutable data

hwwhww opened the pull request:
### What was wrong?
To generalize `VMState` for both sharding client and non-sharding client in the early stage, we need to fix #270: decouple blockchain opcodes and state mutable data.

updated:
As discussion in #270, it makes sense to move `validate_block` from `VMState` to `VM` since `VM.import_block(block)` and  `VM.mine_block()` are in `VM` now.

### How was it fixed?
- [x] Add BlockInfo object to hold the block information opcode values
- [x] Refactor `VMState.block_header` to other format (see #270 for more details)
- [x] Move `validate_block` from `VMState` to `VM`
- [x] Now `VMState` doesn't necessarily need the whole previous 256 headers, the hashes are enough to provide the context. So this PR also changed the `prev_headers` to `prev_hashes` and removed the unused methods.

#### Cute Animal Picture
![levi](https://user-images.githubusercontent.com/9263930/34929669-3943953c-fa00-11e7-8b6b-14459ebc82e8.jpg)

pipermerriam commented on evm/vm/block_info.py:
```diff
@@ -0,0 +1,62 @@
+        self._prev_headers = prev_headers
+
+    @classmethod
+    def generate_by_block_header(cls, block_header, prev_headers, coinbase=None):
```
This would closer align with existing naming conventions if it was `generate_from_block_header` or more concisely, just `from_block_header`

hwwhww replied:
My carelessness, thank you.

pipermerriam commented on evm/vm/block_info.py:
```diff
@@ -0,0 +1,62 @@
+    _number = None
+    _difficulty = None
+    _gas_limit = None
+    _prev_headers = None
```
Given all of these are written as direct passthrough `property` methods with no setters, I think it might be better to remove the underscore prefix and property methods and expect them to be accessed directly.

hwwhww replied:
I think the purpose of this strategy *is* making these fields won't be changed after initialization. In this way, it will raise `AttributeError` when someone tries to modify these properties. But maybe there's a better approach for this purpose? I'd love to hear your thoughts.

pipermerriam replied:
I don't think that these will throw errors if modified.  I believe python will happily let you assign over top of `@property` methods.

- The pragmatic part of me says to just use direct properties and we can have *soft* enforcement of not overwriting values (as in, in code reviews and things, we have a pretty strong *culture* of immutability and well defined boundaries when we require mutability).
- The purist likes this idea and suggests making our own `read_only_property` decorator which includes a `setter` which raises such an attribute error.

I'm leaner towards *soft* enforcement, but I'm willing to hear any arguments you have for the later.

pipermerriam commented on evm/vm/block_info.py:
```diff
@@ -0,0 +1,62 @@
+class BlockInfo():
```
Wondering if it would be better to not use the name *block* and instead to call this something like

* VMContext
* ExecutionContext

I'm partially thinking about https://github.com/ethereum/py-evm/issues/262 and whether the changes in this pull request can help us fix that issue as well (but not in this PR)

hwwhww commented:
I try to pass `VM.block.header.state_root` to initialize `VMState`. Since the `VM.block.block_header` no more be used as a mutable object passed to `VMState.__init__(..., block_header,...)`, it turns out that there is a new side effect: it needs to maintain `VMState.block.header.state_root` whenever modifying the `state_db` in `VM` level. For example:
https://github.com/hwwhww/py-evm/blob/a963f0fdad552f9eade8e1cdcee88dd2d0db8e97/evm/vm/forks/homestead/headers.py#L81-L89
```python
        with vm.state.state_db() as state_db:
            for account in dao_drain_list:
                account = decode_hex(account)
                balance = state_db.get_balance(account)
                state_db.delta_balance(dao_refund_contract, balance)
                state_db.set_balance(account, 0)

+           # Update state_root manually
+           vm.block.header.state_root = state_db.root_hash
```
I'm still investigating any better solution.

hwwhww commented:
#### Updated PR description:
As discussion in #270, it makes sense to move `validate_block` from `VMState` to `VM` since `VM.import_block(block)` and  `VM.mine_block()` are in `VM` now.

#### Changelogs:
1. Moved `validate_block` from `VMState` to `VM`.
2. Now `VMState` doesn't necessarily need the whole previous 256 headers, the hashes are enough to provide the context. So this PR also changed the `prev_headers` to `prev_hashes` and removed the unused methods.

pipermerriam commented on evm/vm/base.py:
```diff
@@ -379,21 +477,21 @@ def get_block_by_header(cls, block_header, db):
-    def get_prev_headers(cls, last_block_hash, db):
-        prev_headers = []
+    def get_prev_hashes(cls, last_block_hash, db):
+        prev_hashes = []
```
Can we use the `to_tuple` decorator here to avoid the `prev_hashes.append` calls.

pipermerriam commented on evm/vm/execution_context.py:
```diff
@@ -0,0 +1,62 @@
+class ExecutionContext():
```
*style nitpick*

You can drop the `()`

pipermerriam commented on evm/vm/forks/frontier/validation.py:
```diff
@@ -18,7 +18,11 @@ def validate_frontier_transaction(vm_state, transaction):
 
-    if vm_state.block_header.gas_used + transaction.gas > vm_state.block_header.gas_limit:
+    if vm_state.receipts:
+        gas_used = vm_state.receipts[-1].gas_used
```
What do you think about adding a`@propery` method for the `vm_state` class since this block of logic is repeated at least one more time elsewhere in the code.

```python
@propert
def gas_used(seelf):
    if vm_state.receipts:
        return ...
    else:
        return 0
```

pipermerriam commented on evm/vm_state.py:
```diff
@@ -141,36 +134,19 @@ def is_key_exists(self, key):
         if (ancestor_depth >= MAX_PREV_HEADER_DEPTH or
                 ancestor_depth < 0 or
-                ancestor_depth >= len(self.prev_headers)):
+                ancestor_depth >= len(self.execution_context.prev_hashes)):
```
We could improve readability here by pulling the full contents of the `if` statement and storing in a variable and then just using the variable in the `if` statement.

```python
is_ancestor_depth_out_of_range = (
    ancestor_depth >= MAX_PREV_HEADER_DEPTH or
    ancestor_depth < 0 or
    ancestor_depth >= len(self.execution_context.prev_hashes)
))
```

hwwhww commented:
#### Changelogs
Refactored `VM` and `VMState` for avoiding to update `state_db` in `VM` level.
1. Moved the reward methods from `VM` to `VMState`.
2. Pick partial logic from `VM.mine_block()` to `VMState.finalize_block(block)` method.

* Now the only one special case of "updating `state_root` manually in VM level” under `evm` is [the DAO fork header initialization](https://github.com/hwwhww/py-evm/blob/18578f1ca25e0a2dc45f003eac9e8f4579be8dc5/evm/vm/forks/homestead/headers.py#L89-L90).
* Under `tests`, I also added 3 lines to update `state_root` manually in VM level for setup the JSON fixtures test cases, but I think it's fine if only for the testing setup.
* With these changes, we also got clear `VM.create_block(block)`.

hwwhww commented:
This branch has passed Travis CI in my repository, so I go :shipit:

lrettig commented on evm/vm/execution_context.py:
```diff
@@ -0,0 +1,62 @@
+class ExecutionContext:
```
Is there a reason this is an old-style python object?

pipermerriam replied:
```python
class AwesomeNewstyleOnlyInPython3:
    ...
```

It's only old-style in python2.  In python 3 this is equivalent to `class Thing(object)`.

lrettig replied:
🤯
